lib/app/features/chat/views/components/messages_list.dart

```diff
@@ -0,0 +1,180 @@
+import 'package:ion/app/features/chat/messages/views/components/message_types/audio_message/audio_message.dart';
+import 'package:ion/app/features/chat/messages/views/components/message_types/emoji_message/emoji_message.dart';
+import 'package:ion/app/features/chat/messages/views/components/message_types/money_message/money_message.dart';
+import 'package:ion/app/features/chat/messages/views/components/message_types/photo_message/photo_message.dart';
+import 'package:ion/app/features/chat/messages/views/components/message_types/poll_message/poll_message.dart';
+import 'package:ion/app/features/chat/messages/views/components/message_types/profile_share_message/profile_share_message.dart';
+import 'package:ion/app/features/chat/messages/views/components/message_types/system_message/system_message.dart';
+import 'package:ion/app/features/chat/messages/views/components/message_types/text_message/text_message.dart';
+import 'package:ion/app/features/chat/messages/views/components/message_types/url_preview_message/url_preview_message.dart';
+import 'package:ion/app/features/chat/messages/views/components/message_types/video_message/video_message.dart';
+import 'package:ion/app/features/chat/model/message_author.dart';
+import 'package:ion/app/features/chat/model/message_list_item.dart';
+import 'package:ion/app/features/chat/model/message_reaction_group.dart';
+
+class ChatMessagesList extends StatelessWidget {
+  const ChatMessagesList(
+    this.messages, {
+    super.key,
+    this.displayAuthorsIncomingMessages = false,
+  });
+
+  final bool displayAuthorsIncomingMessages;
+  final List<MessageListItem> messages;
+
+  bool _isMessageFromDifferentUser(int index, MessageAuthor? currentAuthor) {
+    final message = messages[index];
+    if (message is MessageWithAuthor) {
+      return currentAuthor != (message as MessageWithAuthor).author;
+    }
+    return true;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return ColoredBox(
+      color: context.theme.appColors.primaryBackground,
+      child: ListView.separated(
+        itemCount: messages.length,
+        padding: EdgeInsets.symmetric(vertical: 12.0.s),
+        itemBuilder: (context, index) {
+          final message = messages[index];
+          final isLastMessage = index == (messages.length - 1);
+          final isFirstMessage = index <= 0;
+
+          final author =
+              message is MessageWithAuthor ? (message as MessageWithAuthor).author : null;
+
+          final isLastMessageFromSender =
+              isLastMessage || _isMessageFromDifferentUser(index + 1, author);
+
+          MessageAuthor? authorToDisplay;
+          if (displayAuthorsIncomingMessages) {
+            final isFirstMessageFromSender =
+                isFirstMessage || _isMessageFromDifferentUser(index - 1, author);
+            authorToDisplay = isFirstMessageFromSender ? author : null;
```

Comment:
can we move a calculation of the logic related to the `authorToDisplay` into the provider?

Reply:
Yep, moved.